rtengine/demosaic_algos.cc

```diff
@@ -62,88 +62,100 @@ extern const Settings* settings;
+                avg[j][h] = avgL;
+                dev[j][h] = std::max(0.001f, SQR(temp[j - 4][h] - avgL) + SQR(temp[j - 3][h] - avgL) + SQR(temp[j - 2][h] - avgL) + SQR(temp[j - 1][h] - avgL) + SQR(temp[j][h] - avgL) + SQR(temp[j + 1][h] - avgL) + SQR(temp[j + 2][h] - avgL) + SQR(temp[j + 3][h] - avgL) + SQR(temp[j + 4][h] - avgL));
+            }
+        }
 
-            if (devL < 0.001) {
-                devL = 0.001;
+        for (int j = 5; j < H - 5; j++) {
+            #pragma omp simd
+            for(int h = 0; h < numCols; ++h) {
+                const float avgL = avg[j - 1][h];
+                const float avgR = avg[j + 1][h];
+                const float devL = dev[j - 1][h];
+                const float devR = dev[j + 1][h];
+                hpmap[j][k + h] = avgL + (avgR - avgL) * devL / (devL + devR);
             }
+        }
+    }
+    for (; k < col_to; k++) {
+        for (int i = 5; i < H - 5; i++) {
+            temp[i][0] = std::fabs((rawData[i - 5][k] - rawData[i + 5][k]) - 8 * (rawData[i - 4][k] - rawData[i + 4][k]) + 27 * (rawData[i - 3][k] - rawData[i + 3][k]) - 48 * (rawData[i - 2][k] - rawData[i + 2][k]) + 42 * (rawData[i - 1][k] -rawData[i + 1][k]));
+        }
 
-            dev[j] = devL;
+        for (int j = 4; j < H - 4; j++) {
+            const float avgL = (temp[j - 4][0] + temp[j - 3][0] + temp[j - 2][0] + temp[j - 1][0] + temp[j][0] + temp[j + 1][0] + temp[j + 2][0] + temp[j + 3][0] + temp[j + 4][0]) / 9.f;
+            avg[j][0] = avgL;
+            dev[j][0] = std::max(0.001f, SQR(temp[j - 4][0] - avgL) + SQR(temp[j - 3][0] - avgL) + SQR(temp[j - 2][0] - avgL) + SQR(temp[j - 1][0] - avgL) + SQR(temp[j][0] - avgL) + SQR(temp[j + 1][0] - avgL) + SQR(temp[j + 2][0] - avgL) + SQR(temp[j + 3][0] - avgL) + SQR(temp[j + 4][0] - avgL));
         }
 
         for (int j = 5; j < H - 5; j++) {
-            float avgL = avg[j - 1];
-            float avgR = avg[j + 1];
-            float devL = dev[j - 1];
-            float devR = dev[j + 1];
+            const float avgL = avg[j - 1][0];
+            const float avgR = avg[j + 1][0];
+            const float devL = dev[j - 1][0];
+            const float devR = dev[j + 1][0];
             hpmap[j][k] = avgL + (avgR - avgL) * devL / (devL + devR);
         }
     }
 
-    delete [] temp;
-    delete [] avg;
-    delete [] dev;
 }
 
 //%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
 void RawImageSource::hphd_horizontal (float** hpmap, int row_from, int row_to)
 {
-    float* temp = new float[max(W, H)];
-    float* avg = new float[max(W, H)];
-    float* dev = new float[max(W, H)];
+//    BENCHFUN
+    float* temp = new float[W];
+    float* avg = new float[W];
+    float* dev = new float[W];
 
-    memset (temp, 0, max(W, H)*sizeof(float));
-    memset (avg, 0, max(W, H)*sizeof(float));
-    memset (dev, 0, max(W, H)*sizeof(float));
+    memset(temp, 0, W * sizeof(float));
+    memset(avg, 0, W * sizeof(float));
+    memset(dev, 0, W * sizeof(float));
```

Comment:
@heckflosse I came across an interesting [detail](https://en.cppreference.com/w/cpp/language/new) (under *Construction*):

> * If type is an array type, an array of objects is initialized.
>   * If initializer is an empty pair of parentheses, each element is value-initialized.

As `new()` under Linux apparently zero-initialized by default, I had a look at the assembly:

```
        int* const test = new int[items]();
  400896:       bf 00 00 04 00          mov    $0x40000,%edi
  40089b:       e8 70 fe ff ff          callq  400710 <operator new[](unsigned long)@plt>
  4008a0:       48 89 c1                mov    %rax,%rcx
  4008a3:       48 89 c8                mov    %rcx,%rax
  4008a6:       ba ff ff 00 00          mov    $0xffff,%edx
  4008ab:       eb 0e                   jmp    4008bb <main+0x35>
  4008ad:       c7 00 00 00 00 00       movl   $0x0,(%rax)
  4008b3:       48 83 c0 04             add    $0x4,%rax
  4008b7:       48 83 ea 01             sub    $0x1,%rdx
  4008bb:       48 83 fa ff             cmp    $0xffffffffffffffff,%rdx
  4008bf:       75 ec                   jne    4008ad <main+0x27>
  4008c1:       48 89 4d e8             mov    %rcx,-0x18(%rbp)

       int* const test = new int[items];
  400896:       bf 00 00 04 00          mov    $0x40000,%edi
  40089b:       e8 70 fe ff ff          callq  400710 <operator new[](unsigned long)@plt>
  4008a0:       48 89 45 e8             mov    %rax,-0x18(%rbp)
```

:tada: 

So it looks like you can spare the `memset()`s by simply adding  `()` to `new float[W]`. That will not make the code faster but less verbose.

Best,
Flössie

Reply:
@Floessie Thanks for the review :+1: 
I included your suggested change in the [librtprocess hphd code](https://github.com/CarVac/librtprocess/commit/5889d38a0b977a331f65ea99bf6ae866ce3f95ad#diff-5990c9321c7b70ce2e381ae019e6b06dR119)